arch/x86/core/spec_ctrl.c

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <cpuid.h> /* Header provided by the toolchain. */
+
+#include <init.h>
+#include <kernel_structs.h>
+#include <kernel_arch_data.h>
+#include <kernel_arch_func.h>
+#include <kernel.h>
+
+#define CPUID_EXTENDED_FEATURES_LVL 7
+
+#define CPUID_SPEC_CTRL BIT(26)
+#define SPEC_CTRL_SSBD BIT(2)
+
+static int
+cpu_has_spec_ctrl(void)
+{
+	u32_t eax, ebx, ecx = 0, edx;
+
+	if (!__get_cpuid(CPUID_EXTENDED_FEATURES_LVL,
+			 &eax, &ebx, &ecx, &edx)) {
```

Comment:
Had to dig around to discover that this was a GCC gadget.  :)

Looking at the source, it's actually more complicated than we want as it does probing to see if the instruction is going to work on the current hardware.  We don't care about that, because we've already gated off the code so it only runs on vulnerable CPUs, all of which of course have cpuid.  This should probably just be a one-line inline assembly thing.

Reply:
Not really -- __get_cpuid() will return 1 if that CPUID level is supported.  I don't really know if all CPUs vulnerable to this thing will support CPUID level 7, hence the check.